File: binding-csharp/runtime/src/test/csharp/Etch/Transport/Filter/TestLogger.cs
// $Id$
// 
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements. See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership. The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License. You may obtain a copy of the License at
// 
//   http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied. See the License for the
// specific language governing permissions and limitations
// under the License.
// 
using System;
using Org.Apache.Etch.Bindings.Csharp.Msg;
using Org.Apache.Etch.Bindings.Csharp.Support;
using Org.Apache.Etch.Bindings.Csharp.Util;
using NUnit.Framework;

namespace Org.Apache.Etch.Bindings.Csharp.Transport.Filter
{
    [TestFixture]
    public class TestLogger
    {
        private MyTransport transport;
        private MySession session;
        private MyValueFactory vf;
        private Resources resources;

        [SetUp]
        public void Initialize()
        {
            transport = new MyTransport();
            session = new MySession();
            vf = new MyValueFactory("tcp:");
            resources = new Resources();
            resources.Add(TransportConsts.VALUE_FACTORY, vf);
        }

        private Logger newLogger(string uri) 
	    {
		    return new Logger(transport,new URL(uri),null);
	    }

        [Test]
	    public void constructor1() {
		    String uri = "tcp://localhost:4008?filter=Logger";
		    Logger.testFlag = true;
		    Logger logger = newLogger(uri);
		    Assert.AreEqual(logger.GetFileName(),"Log.txt");
            Assert.AreNotSame(logger.GetFileName(),logger.GetTestRenamedFileName());
	    }

        [Test]
        public void constructor2()
        {
            String uri = "tcp://localhost:4008?filter=Logger&Logger.filePath=Test.txt";
            Logger.testFlag = true;
            Logger logger = newLogger(uri);
            Assert.AreEqual(logger.GetFileName(), "Test.txt");
            Assert.AreNotSame(logger.GetFileName(), logger.GetTestRenamedFileName());
        }

        [Test]
	    public void logger() {
		    String uri = "tcp://localhost:4008?filter=Logger";
		    Logger.testFlag = true;
		    Logger logger = newLogger(uri);
		    logger.SetSession( session );
		    Assert.AreSame(transport.GetSession(),logger);
		    Assert.AreSame(logger.GetSession(),session);
	    }

        [Test]
        public void logger1()
        {
            String uri = "tcp://localhost:4008?filter=Logger";
            Logger.testFlag = true;
            Logger logger = newLogger(uri);
            logger.SetSession(session);
            Assert.AreSame(transport.GetSession(), logger);
            Assert.AreSame(logger.GetSession(), session);
            Message m = constructMessage();
            logger.TransportMessage(null,m);
            Assert.AreEqual(m.ToString(),transport.msg.ToString());
        }

        [Test]
        public void logger2()
        {
            String uri = "tcp://localhost:4008?filter=Logger";
            Logger.testFlag = true;
            Logger logger = newLogger(uri);
            logger.SetSession(session);
            Assert.AreSame(transport.GetSession(), logger);
            Assert.AreSame(logger.GetSession(), session);
            Message m = constructMessage();
            logger.SessionMessage(null, m);
            Assert.AreEqual(m.ToString(), session.msg.ToString());
        }

        private Message constructMessage()
	    {	
		    XType mt_request = new XType( "_Etch_Logger_Test" );
		
		
		    XType mt_response = new XType( "_Etch_Logger_Test_Resp" );
			
		    mt_request.SetResult( mt_response );
		
		    vf.AddType( mt_request);
		
		    Message msg = new Message( mt_request, vf );
	
		    return msg;
	}


        public enum What
        {
            TRANSPORT_MESSAGE,
            SESSION_MESSAGE
        }

        static Object cook(Object o)
        {
            if (o is Exception)
            {
                Exception e = (Exception)o;
                e.GetBaseException();
                throw e;
            }
            return o;
        }

        public class MyTransport : TransportMessage
        {

            public What what;

            public Who recipient;

            public Message msg;

            public Object is_server;

            private SessionMessage session;


            #region TransportMessage Members

            public void TransportMessage(Who recipient, Org.Apache.Etch.Bindings.Csharp.Msg.Message msg)
            {
                this.what = What.TRANSPORT_MESSAGE;
                this.recipient = recipient;
                this.msg = msg;
            }

            #endregion

            #region Transport<SessionMessage> Members

            public object TransportQuery(object query)
            {
                if ((string)query == TransportConsts.IS_SERVER)
                    return cook(is_server);
                throw new Exception("The method or operation is not implemented.");
            }

            public void TransportControl(object control, object value)
            {
                throw new Exception("The method or operation is not implemented.");
            }

            public void TransportNotify(object eventObj)
            {
                throw new Exception("The method or operation is not implemented.");
            }

            public void SetSession(SessionMessage session)
            {
                this.session = session;
            }

            public SessionMessage GetSession()
            {
                return session;
            }

            #endregion
        }

        public class MySession : SessionMessage
        {


            public What what;

            public Who sender;

            public Message msg;

            public bool handled;

            public Object user;

            public Object password;

            public Boolean up;



            #region SessionMessage Members

            public bool SessionMessage(Who sender, Org.Apache.Etch.Bindings.Csharp.Msg.Message msg)
            {
                this.what = What.SESSION_MESSAGE;
                this.sender = sender;
                this.msg = msg;
                return handled;
            }

            #endregion

            #region Session Members

            public object SessionQuery(object query)
            {
                return query;
            }

            public void SessionControl(object control, object value)
            {
                throw new Exception("The method or operation is not implemented.");
            }

            public void SessionNotify(object eventObj)
            {
                if (eventObj.Equals(SessionConsts.UP))
                {
                    up = true;
                    return;
                }

                if (eventObj.Equals(SessionConsts.DOWN))
                {
                    up = false;
                    return;
                }



                throw new Exception("The method or operation is not implemented.");
            }

            #endregion
        }

        public class MyValueFactory : DefaultValueFactory
        {

            public MyValueFactory(String uri)
                : base(uri, types, class2type)
            {
            }

            private readonly static TypeMap types = new TypeMap();

            private readonly static Class2TypeMap class2type = new Class2TypeMap();

            static MyValueFactory()
            {
                DefaultValueFactory.Init(types, class2type);
            }

        }






    }
}

File: binding-java/runtime/src/test/java/org/apache/etch/bindings/java/transport/filters/TestLogger.java
/* $Id$
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.etch.bindings.java.transport.filters;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotSame;
import static org.junit.Assert.assertSame;

import org.apache.etch.bindings.java.msg.Message;
import org.apache.etch.bindings.java.msg.Type;
import org.apache.etch.bindings.java.msg.TypeMap;
import org.apache.etch.bindings.java.support.Class2TypeMap;
import org.apache.etch.bindings.java.support.DefaultValueFactory;
import org.apache.etch.bindings.java.transport.SessionMessage;
import org.apache.etch.bindings.java.transport.TransportMessage;
import org.apache.etch.util.Log;
import org.apache.etch.util.Resources;
import org.apache.etch.util.URL;
import org.apache.etch.util.core.Who;
import org.apache.etch.util.core.io.Session;
import org.apache.etch.util.core.io.Transport;
import org.junit.Test;


/** */
public class TestLogger
{
	
	private final MyTransport transport = new MyTransport();
	private final MySession session = new MySession();
	private final MyValueFactory vf = new MyValueFactory("tcp:" );
	private final Resources resources = new Resources();
	{
		resources.put( Transport.VALUE_FACTORY, vf );
		Log.addSink( null );
	}

	
	private Logger newLogger(String uri) throws Exception
	{
		return new Logger(transport,new URL(uri),null);
	}
	
	/** @throws Exception */
	@Test
	public void constructor1() throws Exception {
		String uri = "tcp://localhost:4802?filter=Logger";
		Logger.testFlag = true;
		Logger logger = newLogger(uri);
		assertEquals(logger.getFileName(),"Log.txt");
		assertNotSame( logger.getFileName(),logger.getTestRenamedFileName());
		
	}
	
	/** @throws Exception */
	@Test
	public void constructor2() throws Exception {
		String uri = "tcp://localhost:4802?filter=Logger&Logger.filePath=Test.txt";
		Logger.testFlag = true;
		Logger logger = newLogger(uri);
		assertEquals(logger.getFileName(),"Test.txt");
		assertNotSame( logger.getFileName(),logger.getTestRenamedFileName());
	}
	
	/** @throws Exception */
	@Test
	public void logger() throws Exception {
		String uri = "tcp://localhost:4802?filter=Logger";
		Logger.testFlag = true;
		Logger logger = newLogger(uri);
		logger.setSession( session );
		assertSame(transport.getSession(),logger);
		assertSame(logger.getSession(),session);
	}
	
	/** @throws Exception */
	@Test
	public void logger1() throws Exception {
		String uri = "tcp://localhost:4802?filter=Logger";
		Logger.testFlag = true;
		Logger logger = newLogger(uri);
		logger.setSession( session );
		assertSame(transport.getSession(),logger);
		assertSame(logger.getSession(),session);
		Message m = constructMessage();
		logger.transportMessage( null, m );
		assertEquals(m.toString(),transport.msg.toString());
	}
	
	/** @throws Exception */
	@Test
	public void logger2() throws Exception {
		String uri = "tcp://localhost:4802?filter=Logger";
		Logger.testFlag = true;
		Logger logger = newLogger(uri);
		logger.setSession( session );
		assertSame(transport.getSession(),logger);
		assertSame(logger.getSession(),session);
		Message m = constructMessage();
		logger.sessionMessage( null, m );
		assertEquals(m.toString(),session.msg.toString());
	}
	
	
	
	private Message constructMessage() throws Exception
	{
		
		Type mt_request = new Type( "_Etch_Logger_Test" );
		
		
		Type mt_response = new Type( "_Etch_Logger_Test_Resp" );
			
		mt_request.setResult( mt_response );
		
		vf.addType( mt_request);
		
		Message msg = new Message( mt_request, vf );
	
		return msg;
	}

	/** */
	enum What
	{
		/** */ TRANSPORT_MESSAGE,
		/** */ SESSION_MESSAGE
	}

	@SuppressWarnings("hiding")
	private static class MyTransport implements TransportMessage
	{
		/** */
		public MyTransport()
		{
			// nothing to do.
		}

		/** */ public What what;
		/** */ public Who recipient;
		/** */ public Message msg;
		
		public void transportMessage( Who recipient, Message msg )
			throws Exception
		{
			Log.report( "transportMessage", "recipient", recipient, "msg", msg );
			what = What.TRANSPORT_MESSAGE;
			this.recipient = recipient;
			
			this.msg = msg;
		}

		public SessionMessage getSession()
		{
			return session;
		}

		public void setSession( SessionMessage session )
		{
			this.session = session;
		}
		
		private SessionMessage session;

		public Object transportQuery( Object query ) throws Exception
		{
			
			throw new UnsupportedOperationException( "unknown query "+query );
		}

		public void transportControl( Object control, Object value )
			throws Exception
		{
			
			
			throw new UnsupportedOperationException( "unknown control "+control );
		}

		public void transportNotify( Object event ) throws Exception
		{
			throw new UnsupportedOperationException( "unknown notify "+event );
		}
	}
	
	@SuppressWarnings("hiding")
	private static class MySession implements SessionMessage
	{
		/** */
		public MySession()
		{
			// nothing to do.
		}

		/** */ public What what;
		/** */ public Who sender;
		/** */ public Message msg;
		/** */ public Boolean up;
		
		
		public boolean sessionMessage( Who sender, Message msg )
			throws Exception
		{
			Log.report( "sessionMessage", "sender", sender, "msg", msg );
			what = What.SESSION_MESSAGE;
			this.sender = sender;
			this.msg = msg;
			return true;
		}

		public Object sessionQuery( Object query ) throws Exception
		{
			return query;
		}

		public void sessionControl( Object control, Object value )
			throws Exception
		{
			Log.report( "sessionControl", "control", control, "value", value );
			throw new UnsupportedOperationException( "unknown control "+control );
		}

		public void sessionNotify( Object event ) throws Exception
		{
			if (event == Session.UP)
			{
				up = true;
				return;
			}
			
			if (event == Session.DOWN)
			{
				up = false;
				return;
			}
		}
	}
	
	private static class MyValueFactory extends DefaultValueFactory
	{
		/**
		 * @param uri
		 */
		public MyValueFactory( String uri )
		{
			super( uri, types, class2type );
		}
		
		private final static TypeMap types = new TypeMap();
		
		private final static Class2TypeMap class2type = new Class2TypeMap();

		static
		{
			DefaultValueFactory.init( types, class2type );
		}
	}
	
}
